File: tests/test_pointcloud_stream.mjs
import assert from 'node:assert/strict';
import { readFileSync } from 'node:fs';
import test from 'node:test';
import { runInNewContext } from 'node:vm';

const streamSource = readFileSync(new URL('../robot_dashboard/static/pointcloud_stream.js', import.meta.url), 'utf8');
const appSource = readFileSync(new URL('../robot_dashboard/static/app.js', import.meta.url), 'utf8');

const sandbox = {
  ArrayBuffer,
  DataView,
  Float32Array,
  TextDecoder,
  Uint8Array,
};
sandbox.globalThis = sandbox;
sandbox.window = sandbox;
runInNewContext(streamSource, sandbox);
const stream = sandbox.RobotPointCloudStream;

function encodeFrame(metadata, values) {
  const json = new TextEncoder().encode(JSON.stringify({
    ...metadata,
    encoding: 'float32le',
    point_count: values.length / 3,
  }));
  const payloadOffset = (16 + json.byteLength + 3) & ~3;
  const buffer = new ArrayBuffer(payloadOffset + values.length * 4);
  const view = new DataView(buffer);
  for (const [index, value] of [0x52, 0x53, 0x50, 0x43].entries()) view.setUint8(index, value);
  view.setUint8(4, 1);
  view.setUint8(5, 0);
  view.setUint16(6, json.byteLength, true);
  view.setUint32(8, values.length / 3, true);
  view.setUint32(12, values.length * 4, true);
  new Uint8Array(buffer, 16, json.byteLength).set(json);
  new Float32Array(buffer, payloadOffset, values.length).set(values);
  return buffer;
}

test('binary decoder returns aligned float32 XYZ without JSON point expansion', () => {
  const frame = encodeFrame({ seq: 4, stream_id: 'server-A', topic: '/Laser_map' }, [1, 2, 3, -4, 5.5, 6]);
  const decoded = stream.decodeFrame(frame);
  assert.equal(decoded.seq, 4);
  assert.equal(decoded.stream_id, 'server-A');
  assert.equal(decoded.point_count, 2);
  assert.deepEqual(Array.from(decoded.points), [1, 2, 3, -4, 5.5, 6]);
  assert.equal(decoded.points.buffer, frame, 'aligned payload should remain zero-copy');
  assert.equal(decoded.prevalidated, true);
});

test('binary decoder rejects bad magic, bad lengths and oversized claims', () => {
  const frame = encodeFrame({ seq: 1 }, [1, 2, 3]);
  const badMagic = frame.slice(0);
  new Uint8Array(badMagic)[0] = 0;
  assert.throws(() => stream.decodeFrame(badMagic), /magic/);
  assert.throws(() => stream.decodeFrame(frame.slice(0, -1)), /length/);
  const badCount = frame.slice(0);
  new DataView(badCount).setUint32(8, 2_000_000, true);
  assert.throws(() => stream.decodeFrame(badCount), /payload size/);
});

test('registered-cloud reservoir is bounded, idempotent and source-scoped', () => {
  const reservoir = new stream.RegisteredCloudReservoir(1000);
  const values = new Float32Array(1500 * 3);
  for (let index = 0; index < 1500; index += 1) {
    values[index * 3] = index;
    values[index * 3 + 1] = index / 2;
    values[index * 3 + 2] = 1;
  }
  const first = reservoir.ingest({ seq: 1, topic: '/Laser_map', frame_id: 'map', source_points: 1500, points: values }, null);
  assert.equal(first.sent_points, 1000);
  assert.equal(first.points.length, 3000);
  assert.equal(first.display_capped, true);
  assert.equal(reservoir.ingest({ seq: 1, topic: '/Laser_map', frame_id: 'map', points: values }, null), first);

  const nextSource = reservoir.ingest({ seq: 1, topic: '/lidar_points', frame_id: 'hesai', points: Float32Array.of(1, 2, 3) }, 1000);
  assert.equal(nextSource.sent_points, 1);
  assert.equal(nextSource.points.length, 3);
});

test('mapping transport is page-aware, view-aware and resets on server epoch change', () => {
  const activateStart = appSource.indexOf('function activatePage(');
  const activateEnd = appSource.indexOf('\nfunction showToast', activateStart);
  const activate = appSource.slice(activateStart, activateEnd);
  assert.match(activate, /syncPointcloudTransport\(\)/);
  assert.match(activate, /syncCameraTransport\(\)/);
  assert.match(appSource, /function pointcloudTransportWanted\(\)[\s\S]{0,160}activePage === 'mapping'[\s\S]{0,160}desiredMapView\(\) !== 'occupancy'/);
  assert.match(appSource, /streamChanged \|\| legacySequenceRollback/);
  assert.match(appSource, /resetPointcloudStream\(incomingStreamId\)/);
  assert.match(appSource, /document\.addEventListener\('visibilitychange',[\s\S]{0,360}syncPointcloudTransport\(\)/);
});

test('camera relay is requested only while the Sensors page is visible', () => {
  assert.match(appSource, /function cameraTransportWanted\(\)[\s\S]{0,120}activePage === 'sensors' && !document\.hidden/);
  assert.match(appSource, /function disconnectCamera\(\)[\s\S]{0,500}cameraSocketGeneration \+= 1/);
  assert.doesNotMatch(appSource, /initializeRobotProfiles\(\);\s*connectCamera\(\)/);
});

File: tests/test_pointcloud_stream.py
import struct
import unittest

from robot_dashboard.pointcloud_stream import (
    MAX_POINT_BYTES,
    PointCloudFrameError,
    decode_pointcloud_frame,
    encode_pointcloud_frame,
)


class PointCloudStreamCodecTests(unittest.TestCase):
    def test_round_trip_keeps_metadata_and_packed_xyz(self):
        points = struct.pack("<6f", 1.0, -2.0, 3.5, 4.25, 5.0, -6.0)
        frame = encode_pointcloud_frame(
            {
                "seq": 17,
                "stream_id": "test-stream",
                "topic": "/Laser_map",
                "bounds": {"min": [1.0, -2.0, -6.0], "max": [4.25, 5.0, 3.5]},
                "points": [999],
                "points_bytes": b"not-public",
            },
            points,
        )

        metadata, decoded = decode_pointcloud_frame(frame)
        self.assertEqual(decoded, points)
        self.assertEqual(metadata["seq"], 17)
        self.assertEqual(metadata["stream_id"], "test-stream")
        self.assertEqual(metadata["point_count"], 2)
        self.assertEqual(metadata["encoding"], "float32le")
        self.assertTrue(metadata["prevalidated"])
        self.assertNotIn("points", metadata)
        self.assertNotIn("points_bytes", metadata)

    def test_rejects_truncated_or_tampered_frames(self):
        frame = encode_pointcloud_frame({"seq": 1}, struct.pack("<3f", 1, 2, 3))
        for invalid in (
            frame[:10],
            b"NOPE" + frame[4:],
            frame[:-1],
        ):
            with self.subTest(length=len(invalid)):
                with self.assertRaises(PointCloudFrameError):
                    decode_pointcloud_frame(invalid)

        tampered = bytearray(frame)
        tampered[12:16] = (24).to_bytes(4, "little")
        with self.assertRaises(PointCloudFrameError):
            decode_pointcloud_frame(bytes(tampered))

    def test_encoder_enforces_immutable_bounded_xyz_and_finite_json(self):
        with self.assertRaisesRegex(PointCloudFrameError, "immutable"):
            encode_pointcloud_frame({"seq": 1}, bytearray(12))  # type: ignore[arg-type]
        with self.assertRaisesRegex(PointCloudFrameError, "triples"):
            encode_pointcloud_frame({"seq": 1}, b"bad")
        with self.assertRaisesRegex(PointCloudFrameError, "one-million"):
            encode_pointcloud_frame({"seq": 1}, b"\0" * (MAX_POINT_BYTES + 12))
        with self.assertRaisesRegex(PointCloudFrameError, "valid JSON"):
            encode_pointcloud_frame({"seq": 1, "bad": float("nan")}, b"")
        with self.assertRaisesRegex(PointCloudFrameError, "metadata"):
            encode_pointcloud_frame({"seq": 1, "large": "x" * 20_000}, b"")


if __name__ == "__main__":
    unittest.main()
